examples/src/test/kotlin/CodecTest.kt

```diff
@@ -0,0 +1,213 @@
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+internal class CodecTest {
+    // :snippet-start: example-enum
+    enum class PowerStatus {
+        ON,
+        OFF
+    }
+    // :snippet-end:
+
+    // :snippet-start: example-codec
+    class PowerStatusCodec : Codec<PowerStatus> {
+        override fun encode(writer: BsonWriter, value: PowerStatus, encoderContext: EncoderContext) {
+            writer.writeBoolean(value == PowerStatus.ON)
+        }
+
+        override fun decode(reader: BsonReader, decoderContext: DecoderContext): PowerStatus {
+            return when (reader.readBoolean()) {
+                true -> PowerStatus.ON
+                false -> PowerStatus.OFF
+            }
+        }
+
+        override fun getEncoderClass(): Class<PowerStatus> {
+            return PowerStatus::class.java
+        }
+    }
+    // :snippet-end:
+
+    // :snippet-start: example-class
+    data class Monolight(
+        var powerStatus: PowerStatus = PowerStatus.OFF,
+        var colorTemperature: Int? = null
+    ) {
+        override fun toString(): String = "Monolight [powerStatus=$powerStatus, colorTemperature=$colorTemperature]"
+    }
+    // :snippet-end:
+
+    // :snippet-start: example-codec-2
+    class MonolightCodec(registry: CodecRegistry) : Codec<Monolight> {
+        private val powerStatusCodec: Codec<PowerStatus>
+        private val integerCodec: Codec<Int>
+
+        init {
+            powerStatusCodec = registry[PowerStatus::class.java]
+            integerCodec = registry.get(Int::class.java)
+        }
+
+        override fun encode(writer: BsonWriter, value: Monolight, encoderContext: EncoderContext) {
+            writer.writeStartDocument()
+            writer.writeName("powerStatus")
+            powerStatusCodec.encode(writer, value.powerStatus, encoderContext)
+            writer.writeName("colorTemperature")
+            integerCodec.encode(writer, value.colorTemperature, encoderContext)
+            writer.writeEndDocument()
+        }
+
+        override fun decode(reader: BsonReader, decoderContext: DecoderContext): Monolight {
+            val monolight = Monolight()
+            reader.readStartDocument()
+            while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
+                when (reader.readName()) {
+                    "powerStatus" -> monolight.powerStatus = powerStatusCodec.decode(reader, decoderContext)
+                    "colorTemperature" -> monolight.colorTemperature = integerCodec.decode(reader, decoderContext)
+                    "_id" -> reader.readObjectId()
+                }
+            }
+            reader.readEndDocument()
+            return monolight
+        }
+
+        override fun getEncoderClass(): Class<Monolight> {
+            return Monolight::class.java
+        }
+    }
+    // :snippet-end:
+
+    // TODO: is there a more kotlin semantic way to do this?
+    // :snippet-start: codec-provider
+    class MonolightCodecProvider : CodecProvider {
```

Comment:
Add: `@Suppress("UNCHECKED_CAST")` before method - to remove the intellij false positive highlighting.

Reply:
Added before `override fun <T> get(clazz...`